cpp/src/arrow/dataset/file_base.cc

```diff
@@ -450,6 +450,12 @@ Status WriteNextBatch(WriteState& state, const std::shared_ptr<Fragment>& fragme
   ARROW_ASSIGN_OR_RAISE(auto groups, state.write_options.partitioning->Partition(batch));
   batch.reset();  // drop to hopefully conserve memory
 
+  if (static_cast<int>(state.write_options.max_partitions) < static_cast<int>(0)) {
```

Comment:
```suggestion
  if (state.write_options.max_partitions <= 0) {
```

Reply:
There's no need to cast here since max_partitions is already an int. (The other place had to cast since one side was an int and the other side was size_t.)

Reply:
Also, I'd probably put this validation in WriteInternal() below so it gets checked immediately.